Add management of multiple SeekHead Element

gwen-lg opened the pull request:
Some Matroska files have a multiple SeakHead Elements, like a first one who only indicate the position of an SeekHead Element at the end of the file.

gwen-lg committed:
feature(SeekHead): add management of multiple SeekHead Element

Some Matroska files have a multiple SeakHead Elements, like a first one
who only indicate the position of an SeekHead Element at the end
of the file.

hasenbanck commented on src/lib.rs:
```diff
@@ -1807,7 +1807,22 @@ fn parse_seek_head<R: Read + Seek>(
+                            file.seek(SeekFrom::Start(segment_data_offset + seek_entry.offset))?;
+                            let (_element_id, element_data) = next_element(file)?;
+                            assert!(_element_id == ElementId::SeekHead);
+                            //println!("SeekHead : ID : {element_id:?} Data : `{element_data:?}`");
```
Debug code left

gwen-lg replied:
removed :+1:

hasenbanck commented on src/lib.rs:
```diff
@@ -1807,7 +1807,22 @@ fn parse_seek_head<R: Read + Seek>(
+                        if seek_entry.id == ElementId::SeekHead {
+                            file.seek(SeekFrom::Start(segment_data_offset + seek_entry.offset))?;
+                            let (_element_id, element_data) = next_element(file)?;
+                            assert!(_element_id == ElementId::SeekHead);
```
It this error really unrecoverable? Shouldn't we return an error?

gwen-lg replied:
I added the assert during experimentation, to check than the element found is the one expected
Effectively, an error should be more appropriate

hasenbanck commented on src/lib.rs:
```diff
@@ -1807,7 +1807,22 @@ fn parse_seek_head<R: Read + Seek>(
+                            if let ElementData::Location { offset, size } = element_data {
+                                // Push next SeekHead Element
+                                let _old = optional_seek_head.replace((offset, size));
+                                assert!(_old.is_none());
```
~Again, is this really unrecoverable?~

I think we can just use `.expect()` then here.

gwen-lg replied:
It seems to me that shouldn't happen, multiple SeekHead seems to be organized as linked list and result more from an programming issue.
But I supposed an error could be appropriate.

hasenbanck replied:
Yes, this is a hard error then. But no need to assert as I wrote above.

gwen-lg replied:
expect is for `Some(_)` result, and there is no `expect_none` unlike `expect_err` for Result.
I have miss something ?

hasenbanck replied:
You're right. Let's then just check it with "debug_assert" at least, so that release builds don't panic in production because we messed up.

hasenbanck commented on src/lib.rs:
```diff
@@ -1794,11 +1794,11 @@ fn parse_ebml_header<R: Read + Seek>(r: &mut R) -> Result<EbmlHeader> {
     let mut seek_head = HashMap::new();
 
-    if let Some((seek_head_data_offset, seek_head_data_size)) = optional_seek_head {
+    while let Some((seek_head_data_offset, seek_head_data_size)) = optional_seek_head.take() {
```
I don't understand this. The parameter is still an option, why do you loop on it?

gwen-lg replied:
The idea is to loop on all SeekHead Element.
The `take` remove the element of the option, and the `replace` add a new one if found.
If no new SeekHead Element position is found, they are all processed, the loop can end and the function exit.

hasenbanck replied:
Ah, okay, I see the replace now, thank you for the hint. Yeah, that's fine then. No need to assert that the old is empty, you can just use `expect` then with a short message.

gwen-lg replied:
Ok.
It's seem than the methods was added : https://github.com/rust-lang/rust/pull/62596
but I not found when and why it was removed.

gwen-lg replied:
In https://github.com/rust-lang/rust/issues/62633
The option `x.ok_or(()).expect_err(...)` was proposed, what do you think of this ?

hasenbanck replied:
That's fine too. So either `debug_assert` or `x.ok_or(()).expect_err(...)` it.

hasenbanck reviewed (changes requested):
Thank you for the PR. I think a test for this change would be helpful to understand it's implementation better.

hasenbanck commented:
PR looks fine so far. Can you please add some sort of test for this?

gwen-lg commented:
You think of a mkv file with multiple seekhead ?
I'm not sure if I could create a small test file with this particularity.

hasenbanck commented:
I thought this change was needed to support multiple SeekHead elements. So we should be able to either create the binary stream that implements such multiple heads, or create a test file to do so. Implementing this feature without a test case is not a good idea.

gwen-lg committed:
tests: add multi_seekhead.mkv file with parse test

the multi_seekhead.mkv file have a first seekhead at start of segment who
have the information of a second seekhead segement at end of the file.

gwen-lg commented:
I'm able to create a file with two seekhead.
The first one (at start of the file) contain only the information (position) of the second one, who are at the end of the file.
It's not as small as the file simple.mkv, but its size seems reasonable to me (570K).

hasenbanck commented:
Yes, the test file and test itself looks fine. Thanks!
